### ISSUE TITLE
PostProcessorRegistrationDelegate$BeanPostProcessorChecker.postProcessAfterInitialization WARNING after upgrade spring-framework-6.1.0

### ISSUE THREAD

**Issue by leonchen83** [spring-projects/spring-batch#4489]
Please do a quick search on Github issues first, there might be already a duplicate issue for the one you are about to create.
If the bug is trivial, just go ahead and create the issue. Otherwise, please take a few moments and fill in the following sections:

**Bug description**
```
23-11-17 16:19:16.578 WARN [main PostProcessorRegistrationDelegate$BeanPostProcessorChecker.postProcessAfterInitialization:437]Bean 'batch.job.registry' of type [org.springframework.batch.core.configuration.support.MapJobRegistry] is not eligible for getting processed by all BeanPostProcessors (for example: not eligible for auto-proxying). Is this bean getting eagerly injected into a currently created BeanPostProcessor [org.springframework.batch.core.configuration.support.JobRegistryBeanPostProcessor#0]? Check the corresponding BeanPostProcessor declaration and its dependencies.
```

```
```

**Environment**
```
spring-batch : 5.0.3
spring-framework : 6.1.0
```

**Steps to reproduce**
```
    <bean class="org.springframework.batch.core.configuration.support.JobRegistryBeanPostProcessor">
        <property name="jobRegistry" ref="batch.job.registry" />
    </bean>

    <bean id="batch.job.registry" class = "org.springframework.batch.core.configuration.support.MapJobRegistry">
    </bean>
```

**Expected behavior**

no warning

**Minimal Complete Reproducible example**
Please provide a failing test or a [minimal complete verifiable example](https://github.com/spring-projects/spring-batch/blob/main/ISSUE_REPORTING.md) that reproduces the issue.
Bug reports that are reproducible will take priority in resolution over reports that are not reproducible.

**Comment by mp911de** [comment]
Duplicate of https://github.com/spring-projects/spring-data-jpa/issues/3244 and https://github.com/spring-projects/spring-boot/issues/38558.

We're not quite sure where this comes from.

**Comment by hpoettker** [comment]
The sudden attention comes from the changed log level from `INFO` to `WARN`: https://github.com/spring-projects/spring-framework/issues/24092

The warnings are also logged with Java configuration (i.e. without XML) when the [DefaultBatchConfiguration](https://github.com/spring-projects/spring-batch/blob/main/spring-batch-core/src/main/java/org/springframework/batch/core/configuration/support/DefaultBatchConfiguration.java) is used. The factory method for the `JobRegistryBeanPostProcessor` would ideally be static, as well as that for the `MapJobRegistry`, which the bean post processor depends on. But making this change would mean that the method can no longer be overridden which is how the `DefaultBatchConfiguration` is supposed to be extended.

**Comment by laywin** [comment]
when declare a beanPostProcessor in xxAutoConfiguration. there will be a warn log for the xxAutoConfiguration. and the bbpsInCreation is empty. so the log message would like "..injected into a currently created BeanPostProcessor [] ? Check the corresponding BeanPostProcessor declaration and its dependencies.", which looks a little strange.

**Comment by philippn** [comment]
As far as I understand, the `org.springframework.batch.core.configuration.support.JobRegistrySmartInitializingSingleton` was meant as a remediation of this warning. However, the fix is incomplete because `org.springframework.batch.core.configuration.support.DefaultBatchConfiguration` creates a `JobRegistryBeanPostProcessor` regardless of whether a bean of type `org.springframework.batch.core.configuration.support.JobRegistrySmartInitializingSingleton` exists or not.

**Comment by fmbenhassine** [comment]
@philippn There wasn't really a fix for these warnings in 5.1.1. The introduction of `JobRegistrySmartInitializingSingleton` was meant to provide a replacement for the `JobRegistryBeanPostProcessor`.

The proper fix for Java configuration is addressed in #4547, by replacing the BPP with the SIS in the default configuration.

For XML configuration as reported in this, the now deprecated `JobRegistryBeanPostProcessor` should be replaced with a `JobRegistrySmartInitializingSingleton`:

```diff
--<bean class="org.springframework.batch.core.configuration.support.JobRegistryBeanPostProcessor">
++<bean class="org.springframework.batch.core.configuration.support.JobRegistrySmartInitializingSingleton">
        <property name="jobRegistry" ref="batch.job.registry" />
    </bean>
```

**Comment by fmbenhassine** [comment]
As mentioned in my previous comment, the Java configuration has been updated in #4547 to use SIS instead of a BPP (which will be available in the upcoming 5.2.0-M1). For XML configuration, users need to replace the deprecated BPP with the SIS. A deprecation note was added to the docs: https://docs.spring.io/spring-batch/reference/5.2-SNAPSHOT/job/advanced-meta-data.html#jobregistrybeanpostprocessor.
